pyproject.toml

```diff
@@ -38,6 +38,7 @@ dask = "*"
 toml = "*"
 autograd = "1.6.2"
 scipy = "*"
+opt-einsum = "^3.3.0"
```

Comment:
we need to be careful about this due to the danger of bloating our core dependencies. It became an issue recently and we agreed to let MC know if we add any dependencies. Is this absolutely necessary you think?

Reply:
`opt-einsum` is a pretty tiny dependency (~200kb) and itself only depends on numpy? And it's a dependency of jax too (although that's not core). Otherwise I can also rewrite that part using regular numpy einsum, not a big deal.

Reply:
@lei-flex what do you think about adding this dependency?

Reply:
I removed the opt-einsum dependency, using numpy instead